### PR TITLE
Introduce build script best practices chapter in user guide

### PR DESCRIPTION
**Pull request by bmuschko**
### Context

For more information see https://github.com/gradle/dotorg-docs/issues/204. Contains initial content. We still may want to change the organization of the content based on the overall information architecture.

### CODE REVIEW

File: subprojects/docs/src/docs/userguide/authoringMaintainableBuildScripts.adoc [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+[NOTE]
+====
+The third-party link:https://github.com/nebula-plugins/gradle-lint-plugin[Gradle lint plugin] helps with enforcing a desired code style in a build script if you are looking for appropriate link:https://en.wikipedia.org/wiki/Lint_(software)[linting] automation.
+====
+
+[[sec:declare_task_group_and_description]]
+=== Declare the group and description for tasks
+
+Even new users to a build should to be able to find crucial information quickly and effortlessly. In Gradle you can declare a api:org.gradle.api.Task#setGroup(java.lang.String)[group] and a api:org.gradle.api.Task#setDescription(java.lang.String)[description] for any task of the build. The <<sec:listing_tasks,tasks report>> uses the assigned values to organize and render the task for easy discoverability. Assigning a group and description is most helpful for lifecycle tasks.
```
Comment (eriwen) [inline subprojects/docs/src/docs/userguide/authoringMaintainableBuildScripts.adoc]: > Assigning a group and description is most helpful for lifecycle tasks.

I wouldn't imagine users know what a "lifecycle" task is. I recommend we state that this should be done for any task that build users will invoke directly.

File: subprojects/docs/src/docs/userguide/authoringMaintainableBuildScripts.adoc [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+++++
+
+The output of the tasks report reflects the assigned values.
+
+----
+$ gradle tasks
+
+> Task :tasks
+
+...
```
Comment (eriwen) [inline subprojects/docs/src/docs/userguide/authoringMaintainableBuildScripts.adoc]: :o: I reckon folks'll get it even without the ellipses.

File: subprojects/docs/src/docs/userguide/authoringMaintainableBuildScripts.adoc [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+Documentation tasks
+-------------------
+generateDocs - Generates the HTML documentation for this project.
+
+...
+----
+
+[[sec:avoid_use_of_imperative_logic]]
+=== Avoid the use of imperative logic
```
Comment (eriwen) [inline subprojects/docs/src/docs/userguide/authoringMaintainableBuildScripts.adoc]: I think "Avoid imperative logic in scripts" might be a bit clearer. "the use of" is probably not necessary.

File: subprojects/docs/src/docs/userguide/authoringMaintainableBuildScripts.adoc [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+The Gradle runtime does not enforce a specific style for build logic. For that very reason, it's easy to end up with a build script that mixes declarative DSL elements with imperative, procedural code. Let's talk about some concrete examples.
+
+* _Declarative code:_ Built-in, language-agnostic DSL elements (e.g. api:org.gradle.api.Project#dependencies(groovy.lang.Closure)[] or api:org.gradle.api.Project#repositories(groovy.lang.Closure)[]) or DSLs exposed by plugins
+* _Imperative code:_ Conditional logic or very complex task action implementations
+
+The end goal of every build script should be to only contain declarative language elements which makes the code easier to understand and maintain. Imperative logic should live in binary plugins and which in turn is applied to the build script. As a side product, you automatically enable your team to link:https://guides.gradle.org/designing-gradle-plugins/#reusable_logic_should_be_written_as_binary_plugin[reuse the plugin logic in other projects] if you publish the artifact to a binary repository.
```
Comment (eriwen) [inline subprojects/docs/src/docs/userguide/authoringMaintainableBuildScripts.adoc]: I'd say "live in `buildSrc` or binary plugins" and perhaps link to appropriate sections of "Organizing Build Scripts" or whatever chapter it'll be.

File: subprojects/docs/src/docs/userguide/authoringMaintainableBuildScripts.adoc [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+The following sample build shows a negative example of using conditional logic directly in the build script. While this code snippet is small, it is easy to imagine a full-blown build script using numerous procedural statements and the impact it would have on readability and maintainability. By moving the code into a class link:https://guides.gradle.org/testing-gradle-plugins/[testability] also becomes a valid option.
+
+++++
+<sample id="conditionalLogicDont" dir="userguide/bestPractices/conditionalLogic/dont" title="A build script using conditional logic to create a task">
+    <sourcefile file="build.gradle"/>
+</sample>
+++++
+
+Let's compare the build script with the same logic implemented as a binary plugin. The code might look more involved at first but clearly looks more like typical application code. This particular plugin class lives in the <<sec:build_sources,`buildSrc` directory>> which makes it available to the build script automatically.
```
Comment (eriwen) [inline subprojects/docs/src/docs/userguide/authoringMaintainableBuildScripts.adoc]: I like how Oehme says in [his blog post](https://oehme.github.io/use-buildSrc)

> A script is just a plugin without the class/method boilerplate

We probably can ~steal~ re-use a bunch of his words verbatim for this. They're good words.

File: subprojects/docs/src/docs/userguide/authoringMaintainableBuildScripts.adoc [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+    <sourcefile file="build.gradle"/>
+</sample>
+++++
+
+[[sec:avoid_use_of_passwords_in_plain_text]]
+=== Avoid the use of passwords in plain text
+
+Most builds need to consume one or many passwords. The reasons for this need may vary. Some builds need a password for publishing artifacts to a secured binary repository, other builds need a password for downloading binary files. Passwords should always kept safe to prevent fraud. Under no circumstance should you add the password to the build script as property in plain text or declare it in a `gradle.properties`. Those files usually live in a version control repository and can be viewed by anyone that has access to it.
+
+Passwords should be stored in encrypted fashion. At the moment Gradle does not provide a built-in mechanism for encrypting, storing and accessing passwords. A good solution for solving this problem is the link:https://github.com/etiennestuder/gradle-credentials-plugin[Gradle Credentials plugin].
```
Comment (eriwen) [inline subprojects/docs/src/docs/userguide/authoringMaintainableBuildScripts.adoc]: Very good.

File: subprojects/docs/src/docs/userguide/authoringMaintainableBuildScripts.adoc [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+// limitations under the License.
+
+[[authoring_maintainable_build_scripts]]
+== Authoring Maintainable Build Scripts
+
+Gradle build scripts combine the qualities of declarative build logic, expressiveness as well as flexibility and rigidity as needed. As a build script author it is easy to fall into the trap of striking the wrong balance or applying poor coding habits. This chapter describes best practices for writing your build script in a meaningful, yet flexible and efficient way.
+
+[NOTE]
+====
+The third-party link:https://github.com/nebula-plugins/gradle-lint-plugin[Gradle lint plugin] helps with enforcing a desired code style in a build script if you are looking for appropriate link:https://en.wikipedia.org/wiki/Lint_(software)[linting] automation.
```
[the remark this answers is not in the document]
Comment (bmuschko) [reply subprojects/docs/src/docs/userguide/authoringMaintainableBuildScripts.adoc]: done

File: subprojects/docs/src/docs/userguide/authoringMaintainableBuildScripts.adoc (hunk 1 above)
Comment (bmuschko) [reply subprojects/docs/src/docs/userguide/authoringMaintainableBuildScripts.adoc]: done

File: subprojects/docs/src/docs/userguide/authoringMaintainableBuildScripts.adoc (hunk 2 above)
Comment (bmuschko) [reply subprojects/docs/src/docs/userguide/authoringMaintainableBuildScripts.adoc]: done

File: subprojects/docs/src/docs/userguide/authoringMaintainableBuildScripts.adoc (hunk 3 above)
Comment (bmuschko) [reply subprojects/docs/src/docs/userguide/authoringMaintainableBuildScripts.adoc]: done

File: subprojects/docs/src/docs/userguide/authoringMaintainableBuildScripts.adoc (hunk 4 above)
Comment (bmuschko) [reply subprojects/docs/src/docs/userguide/authoringMaintainableBuildScripts.adoc]: The important point I am trying to make here is that you actually need to publish your artifact to be able to reuse it in other project. `buildSrc` and binary plugin won't do per se. The dependency actually needs to be available from another project (that likely lives in a different repo).

File: subprojects/docs/src/docs/userguide/authoringMaintainableBuildScripts.adoc (hunk 5 above)
Comment (bmuschko) [reply subprojects/docs/src/docs/userguide/authoringMaintainableBuildScripts.adoc]: Thanks for the pointer. Wasn't aware of the blog post. I changes some things around.

File: subprojects/docs/src/docs/userguide/authoringMaintainableBuildScripts.adoc (hunk 6 above)
Comment (bmuschko) [reply subprojects/docs/src/docs/userguide/authoringMaintainableBuildScripts.adoc]: Thanks!

File: subprojects/docs/src/docs/userguide/authoringMaintainableBuildScripts.adoc [hunk 8]
```diff
@@ -42,17 +42,34 @@ $ gradle tasks
...
 -------------------
 generateDocs - Generates the HTML documentation for this project.
-
-...
 ----
 
-[[sec:avoid_use_of_imperative_logic]]
-=== Avoid the use of imperative logic
+[[sec:minimize_logic_executed_configuration_phase]]
+=== Minimize logic executed during the configuration phase
```
Comment (eriwen) [inline subprojects/docs/src/docs/userguide/authoringMaintainableBuildScripts.adoc]: Would you say this belongs in the script maintainability doc? It seems covered pretty well in the Performance guide, which is sort of a part of the best practices docs.

File: subprojects/docs/src/docs/userguide/authoringMaintainableBuildScripts.adoc (hunk 4 above)
Comment (eriwen) [reply subprojects/docs/src/docs/userguide/authoringMaintainableBuildScripts.adoc]: Ok

File: subprojects/docs/src/docs/userguide/authoringMaintainableBuildScripts.adoc [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+
+Gradle build scripts combine the qualities of declarative build logic, expressiveness as well as flexibility and rigidity as needed. As a build script author it is easy to fall into the trap of striking the wrong balance or applying poor coding habits. This chapter describes best practices for writing your build script in a meaningful, yet flexible and efficient way.
+
+[NOTE]
+====
+The third-party link:https://github.com/nebula-plugins/gradle-lint-plugin[Gradle lint plugin] helps with enforcing a desired code style in a build script if you are looking for appropriate linting automation.
+====
+
+[[sec:declare_task_group_and_description]]
+=== Declare the group and description for tasks
```
Comment (eriwen) [inline subprojects/docs/src/docs/userguide/authoringMaintainableBuildScripts.adoc]: I wonder whether this really ought to be the first tip. I was hoping for something "cooler".

Comment (bmuschko) [reply subprojects/docs/src/docs/userguide/authoringMaintainableBuildScripts.adoc]: Can you expand on what you'd expect to be "cooler"? In terms of the writing, the content or the features? I moved this to the first position as it is the easiest to approach and apply for users.

File: subprojects/docs/src/docs/userguide/authoringMaintainableBuildScripts.adoc (hunk 8 above)
Comment (bmuschko) [reply subprojects/docs/src/docs/userguide/authoringMaintainableBuildScripts.adoc]: I think this topic is so essential that I wouldn't mind having duplicated coverage in different guides with a slightly different angle.

> Would you say this belongs in the script maintainability doc?

It is somewhat about where logic should live _within_ the script. Especially beginners make this mistake all the time. I don't think these users would necessarily look at the performance guide to figure out the issue as they haven't even understood the build lifecycle. _But_ I'd argue that they'd rather look at a best practices chapter.

I'd say the coverage in the performance guide is a bit abstract and doesn't show a concrete code example: https://guides.gradle.org/performance/#configuration. I don't feel that it is very approachable for beginners.

File: subprojects/docs/src/docs/userguide/authoringMaintainableBuildScripts.adoc (hunk 9 above)
Comment (bmuschko) [reply subprojects/docs/src/docs/userguide/authoringMaintainableBuildScripts.adoc]: Oh, I understand your comment now. You were hoping for a better first _topic_ within the chapter. I don't think I have a strong opinion here.

**Pull request merged**